scripts/imgtool/image.py

```diff
@@ -0,0 +1,164 @@
+        return "<Image version={}, header_size={}, pad={}, payloadlen=0x{:x}>".format(
+                self.version,
+                self.header_size,
+                self.pad,
+                len(self.payload))
+
+    def save(self, path):
+        with open(path, 'wb') as f:
+            f.write(self.payload)
+
+    def check(self):
+        """Perform some sanity checking of the image."""
+        # If there is a header requested, make sure that the image
+        # starts with all zeros.
+        if self.header_size > 0:
+            if any(v != 0 for v in self.payload[0:self.header_size]):
+                raise Exception("Padding requested, but image does not start with zeros")
+
+    def sign(self, key):
+        self.add_header(key)
+
+        tlv = TLV()
+
+        # Note that ecdsa wants to do the hashing itself, which means
+        # we get to hash it twice.
+        sha = hashlib.sha256()
+        sha.update(self.payload)
+        digest = sha.digest()
+
+        tlv.add('SHA256', digest)
+
+        sig = key.sign(self.payload)
+        tlv.add(key.sig_tlv(), sig)
+
+        self.payload += tlv.get()
+
+    def add_header(self, key):
+        """Install the image header.
+
+        The key is needed to know the type of signature, and
+        approximate the size of the signature."""
+
+        flags = IMAGE_F[key.sig_type()]
+        tlvsz = 0
+        tlvsz += TLV_HEADER_SIZE + key.sig_len()
+
+        flags |= IMAGE_F['SHA256']
+        tlvsz += 4 + hashlib.sha256().digest_size
+
+        fmt = ('<' +
+            # type ImageHdr struct {
+            'I' +   # Magic uint32
+            'H' +   # TlvSz uint16
+            'B' +   # KeyId uint8
+            'B' +   # Pad1  uint8
+            'H' +   # HdrSz uint16
+            'H' +   # Pad2  uint16
+            'I' +   # ImgSz uint32
+            'I' +   # Flags uint32
+            'BBHI' + # Vers  ImageVersion
+            'I'     # Pad3  uint32
+            ) # }
+        assert struct.calcsize(fmt) == IMAGE_HEADER_SIZE
+        header = struct.pack(fmt,
+                IMAGE_MAGIC,
+                tlvsz, # TlvSz
+                0, # KeyId (TODO: allow other ids)
+                0,  # Pad1
+                self.header_size,
+                0, # Pad2
+                len(self.payload) - self.header_size, # ImageSz
+                flags, # Flags
+                self.version.major,
+                self.version.minor or 0,
+                self.version.revision or 0,
+                self.version.build or 0,
+                0) # Pad3
+        self.payload = bytearray(self.payload)
+        self.payload[:len(header)] = header
+
+    def pad_to(self, size, align):
+        """Pad the image to the given size, with the given flash alignment."""
+        tsize = trailer_sizes[align]
+        padding = size - (len(self.payload) + tsize)
+        if padding < 0:
+            msg = "Image size (0x{:x}) + trailer (0x{:x}) exceeds requested size 0x{:x}".format(
```

Comment:
As discussed off GitHub, calling this the "image size" is misleading, since it has the signature and SHA256 digest TLVs appended already. People can mistake this with the size of their input image (which is what I did at first), especially if they are not familiar with the internals of the mcuboot image format.

Can this error message get cleaned up a bit to make it more clear what happened?